src/box/applier.cc

```diff
@@ -1536,15 +1536,21 @@ applier_synchro_filter_tx(struct stailq *rows)
 	 */
```

Comment:
Can you apply this to the first commit?
```Diff
@@ -999,15 +999,16 @@ wal_assign_lsn(struct vclock *vclock_diff, struct vclock *base,
 		 * Use lsn of the first global row as
 		 * transaction id.
 		 */
-		if ((*row)->group_id != GROUP_LOCAL && tsn == 0) {
+		if (tsn == 0 && (*row)->group_id != GROUP_LOCAL) {
 			tsn = (*row)->lsn;
 			/*
 			 * Remember the tail being processed.
 			 */
 			first_glob_row = row;
 		}
-		(*row)->tsn = tsn == 0 ? (*start)->lsn : tsn;
 	}
+	if (tsn == 0)
+		tsn = (*start)->lsn;
 
 	/*
```
Now you set `row->tsn` 2 times for each row. The diff suggests to only find the target value in the first loop, fallback to `start->tsn` in a branch, and then set all in the second loop (which is already present below this diff). It should help to avoid 1 write for each row.

Also I swapped the condition parts - it is more likely to fail on `tsn == 0`, should help to avoid the second check most of the times.

Reply:
> Now you set row->tsn 2 times for each row. The diff suggests to only find the target value in the first loop, fallback to start->tsn in a branch, and then set all in the second loop (which is already present below this diff). It should help to avoid 1 write for each row.

I don't understand. Why?

In the first loop (`for (struct xrow_header **row = start; row <= end; row++) {`) I set a valid tsn for all rows after the first glob row. And all the local rows before that get `start->tsn`.

In the second loop (`for (struct xrow_header **row = start; row < first_glob_row; row++)`) I only set tsns to the first couple of local rows, which came before first glob row. Not to all rows.

So if you hae a transaction [l, l, ... l, g, ...] where `l` stands for a local row and `g` stands for a global row, the rows up to the first `g` will receive a `(*start)->tsn`, while all the rows after the first `g` will receive a valid tsn. The second loop will rewrite tsns of the rows up to the first `g`,  but not of all the rows.

If I understood you correctly, you want to assign all tsns in the second loop. But then I'll always have to go through two loops. Even if tsn is known right in the first transaction row. And even if there are no global rows in the transaction.

With the current approach both cases above require a single loop (the first one), and the second one is skipped. Meaning there is only one tsn assignment per row. The only case when we write tsn twice happens only for the first local rows of a transaction, up to the first global row.

> Also I swapped the condition parts - it is more likely to fail on tsn == 0, should help to avoid the second check most of the times.

Applied:
```diff
diff --git a/src/box/wal.c b/src/box/wal.c
index ac30de209a..13ed7a0184 100644
--- a/src/box/wal.c
+++ b/src/box/wal.c
@@ -999,7 +999,7 @@ wal_assign_lsn(struct vclock *vclock_diff, struct vclock *base,
                 * Use lsn of the first global row as
                 * transaction id.
                 */
-               if ((*row)->group_id != GROUP_LOCAL && tsn == 0) {
+               if (tsn == 0 && (*row)->group_id != GROUP_LOCAL) {
                        tsn = (*row)->lsn;
                        /*
                         * Remember the tail being processed.
```

Reply:
Sorry, I made a mistake, all is much simpler. I just didn't notice that the 2 loops have different end condition.